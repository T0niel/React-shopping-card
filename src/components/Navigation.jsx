import { ShoppingBasket } from 'lucide-react';
import { useState } from 'react';
import { Link, useLocation } from 'react-router-dom';
import PropTypes from 'prop-types';

const Navigation = ({ shoppingOnClick, cartAmount, links }) => {
  const [shopIconColor, setShopIconColor] = useState('black');
  const location = useLocation();
  console.log(location.pathname);

  return (
    <nav className="h-16 mobile:pr-12 mobile:pl-12 pr-5 pl-5 pt-4 pb-4 font-sans flex">
      <ul className="flex gap-2 text-gray-700 font-medium">
        {links.map(({ id, to, name }) => (
          <li key={id}>
            <Link
              to={location.pathname.slice(1) !== to ? to : undefined}
              className="hover:text-green-400 transition ease-in-out duration-300"
            >
              {name}
            </Link>
          </li>
        ))}
      </ul>
      <div className="ml-auto flex gap-2 items-center">
        <button
          aria-label="shopping card"
          onClick={shoppingOnClick}
          onMouseEnter={() => setShopIconColor('green')}
          onMouseLeave={() => setShopIconColor('black')}
        >
          <ShoppingBasket stroke={shopIconColor}></ShoppingBasket>
        </button>
        <p className="font-mono text-gray-600 text-[18px]">{cartAmount}</p>
      </div>
    </nav>
  );
};

Navigation.propTypes = {
  shoppingOnClick: PropTypes.func.isRequired,
  cartAmount: PropTypes.number,
  links: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      to: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
    })
  ).isRequired,
};

Navigation.defaultProps = {
  cartAmount: 0,
  links: [],
};

export default Navigation;
