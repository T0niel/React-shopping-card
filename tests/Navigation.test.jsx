import { describe, expect, it, vi } from 'vitest';
import Navigation from '../src/components/Navigation';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { BrowserRouter } from 'react-router-dom';
import { v1 as uuid } from 'uuid';

describe('Navigation', () => {
  it('matches snapshot', () => {
    const { container } = render(
      <BrowserRouter>
        <Navigation
          cartAmount={0}
          shoppingOnClick={() => {}}
          links={[{ id: 123, to: '/', name: 'test' }]}
        />
      </BrowserRouter>
    );

    expect(container).toMatchSnapshot();
  });
  it('has a button with shopping card label text for carts', () => {
    render(<Navigation cartAmount={0} shoppingOnClick={() => {}} links={[]} />);

    expect(screen.queryByLabelText('shopping card')).not.toEqual(null);
  });

  it('calls shopping onclick prop when shopping cart clicked', async () => {
    const user = userEvent.setup();
    const onClick = vi.fn();

    render(<Navigation cartAmount={0} shoppingOnClick={onClick} links={[]} />);
    const btn = screen.getByLabelText('shopping card');

    await user.click(btn);
    expect(onClick.mock.calls.length).toEqual(1);

    await user.click(btn);
    expect(onClick.mock.calls.length).toEqual(2);

    await user.click(btn);
    expect(onClick.mock.calls.length).toEqual(3);
  });

  it('does not call shopping onclick prop when shopping cart is not clicked', () => {
    const onClick = vi.fn();
    render(<Navigation cartAmount={0} shoppingOnClick={onClick} links={[]} />);

    expect(onClick.mock.calls.length).toEqual(0);
  });

  it('adds link', () => {
    const testName = uuid();
    render(
      <BrowserRouter>
        <Navigation
          cartAmount={0}
          shoppingOnClick={() => {}}
          links={[{ id: 123, to: '/', name: testName }]}
        />
      </BrowserRouter>
    );

    const link = screen.findByText(testName);
    expect(link).not.equal(null);
  })
});
