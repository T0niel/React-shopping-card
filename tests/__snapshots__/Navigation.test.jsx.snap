// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Navigation > matches snapshot 1`] = `
<div>
  <nav
    class="h-16 pr-12 pl-12 pt-4 pb-4 font-sans flex"
  >
    <ul
      class="flex gap-2 text-gray-700 font-medium"
    >
      <li>
        <a
          class="hover:text-green-400 transition ease-in-out duration-300"
          href="/"
        >
          test
        </a>
      </li>
    </ul>
    <div
      class="ml-auto flex gap-2 items-center"
    >
      <button
        aria-label="shopping card"
      >
        <svg
          class="lucide lucide-shopping-basket"
          fill="none"
          height="24"
          stroke="black"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m15 11-1 9"
          />
          <path
            d="m19 11-4-7"
          />
          <path
            d="M2 11h20"
          />
          <path
            d="m3.5 11 1.6 7.4a2 2 0 0 0 2 1.6h9.8a2 2 0 0 0 2-1.6l1.7-7.4"
          />
          <path
            d="M4.5 15.5h15"
          />
          <path
            d="m5 11 4-7"
          />
          <path
            d="m9 11 1 9"
          />
        </svg>
      </button>
      <p
        class="font-mono text-gray-600 text-[18px]"
      >
        0
      </p>
    </div>
  </nav>
</div>
`;
